analysis/bach_open_taxonomy/bach_open_taxonomy/modelhub/aggregate.py

```diff
@@ -32,74 +55,89 @@ def _generic_aggregation(self, time_aggregation, column, filter, name):
+                                       str,
+                                       Series,
+                                       NotSet] = not_set) -> 'SeriesInt64':
         """
         Calculate the unique sessions in the ObjectivFrame.
 
-        Use any template for aggreation from: https://www.postgresql.org/docs/14/functions-formatting.html
-        ie. ``time_aggregation=='YYYY-MM-DD'`` aggregates by date.
-
-        :param time_aggregation: if None, it uses the time_aggregation set in ObjectivFrame.
         :param filter: the output of this model is only based on the rows for which the filter is True.
+        :param groupby: sets the column(s) to group by.
+            - if `not_set` it defaults to using :py:attr:`ObjectivFrame.model_hub.time_agg`.
+            - if None it aggregates over all data.
         :returns: series with results.
         """
+        groupby = [self._df.mh.time_agg()] if groupby is not_set else groupby
 
-        return self._generic_aggregation(time_aggregation=time_aggregation,
+        return self._generic_aggregation(groupby=groupby,
                                          column='session_id',
                                          filter=filter,
                                          name='unique_sessions')
 
-    def session_duration(self, time_aggregation: str = None):
+    def session_duration(self,
+                         groupby: Union[List[Union[str, Series]],
+                                        str,
+                                        Series,
+                                        NotSet] = not_set) -> 'SeriesInt64':
         """
-        Calculate the average duration of sessions per time interval as set with ``time_aggregation``.
+        Calculate the average duration of sessions.
 
-        :param time_aggregation: if None, it uses the time_aggregation set in ObjectivFrame.
+        :param groupby: sets the column(s) to group by.
+            - if `not_set` it defaults to using :py:attr:`ObjectivFrame.model_hub.time_agg`.
+            - if None it aggregates over all data.
         :returns: series with results.
         """
-        df = self._df.copy_override()
-        df['moment2'] = df.moment
-        if not time_aggregation:
-            time_aggregation = self._df._time_aggregation
-        gb = ['session_id']
-        if time_aggregation:
-            gb = ['session_id', df.moment.dt.sql_format(time_aggregation)]
-
-        session_duration = df.groupby(gb).aggregate({'moment2': ['min', 'max']})
-        session_duration['session_duration'] = session_duration['moment2_max']-session_duration['moment2_min']
+        groupby = [self._df.mh.time_agg()] if groupby is not_set else groupby
+
+        new_groupby = [] if groupby is None else groupby
+        new_groupby = new_groupby if isinstance(groupby, list) else [groupby]
+
+        new_groupby = new_groupby + [self._df.session_id.copy_override(name='_session_id')]  # type: ignore
```

Comment:
I like inline ifs, but sometimes they become a bit difficult to keep up. Maybe?
```suggestion
        new_groupby = [self._df.all_series['session_id'].copy_override(name='_session_id')]]
        
        if groupby is not_set:
              new_groupby.append(self._df.mh.time_agg())
        elif groupby:
              new_groupby += groupby if isinstance(groupby, list) else [groupby]
```

Reply:
it depends on `self._df.all_series['session_id'].copy_override(name='_session_id')` being the last in the list. also the None should be explicitly used as this means aggregating all data (see conversation above). I tried to make it a bit more clear in a new commit.